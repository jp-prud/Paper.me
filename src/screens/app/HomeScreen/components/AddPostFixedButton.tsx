import {ViewStyle} from 'react-native';

import {useNavigation} from '@react-navigation/native';

import {Icon, TouchableOpacityBox, TouchableOpacityBoxProps} from '@components';
import {$shadowProps} from '@theme';

export function AddPostFixedButton() {
  const {navigate} = useNavigation();

  function handlePressNavigate() {
    navigate('PreviewPostScreen');
  }

  return (
    <TouchableOpacityBox
      testID="add-post-fixed-button"
      onPress={handlePressNavigate}
      {...$touchableContainerStyles}
      style={[$shadowProps, $fixedContainerStyle]}>
      <Icon name="pencil" color="white" />
    </TouchableOpacityBox>
  );
}

const $touchableContainerStyles: TouchableOpacityBoxProps = {
  backgroundColor: 'primary',
  alignItems: 'center',
  justifyContent: 'center',
  width: 56,
  height: 56,
  borderRadius: 's32',
  borderWidth: 2,
  borderColor: 'teal200',
};

const $fixedContainerStyle: ViewStyle = {
  position: 'absolute',
  right: 24,
  bottom: 24,
};
