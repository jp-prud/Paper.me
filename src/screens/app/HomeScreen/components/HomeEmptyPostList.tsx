import {useNavigation} from '@react-navigation/native';

import {Box, Button, Text} from '@components';

export function HomeEmptyPostList() {
  const {navigate} = useNavigation();

  function handleCreatePost() {
    navigate('PreviewPostScreen');
  }

  return (
    <Box testID="home-empty-post-list">
      <Text textAlign="center">Nenhum post encontrado.</Text>
      <Box>
        <Text textAlign="center">Seja o primeiro a criar um post.</Text>
        <Button text="Criar post" onPress={handleCreatePost} />
      </Box>
    </Box>
  );
}
