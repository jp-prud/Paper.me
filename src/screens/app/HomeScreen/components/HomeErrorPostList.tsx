import {Box, Button, Text} from '@components';

interface HomeErrorPostListProps {
  refetch: () => void;
}

export function HomeErrorPostList({refetch}: HomeErrorPostListProps) {
  return (
    <Box flex={1} justifyContent="center" testID="home-error-post-list">
      <Box alignItems="center" gap="s4" mb="s24">
        <Text preset="paragraphLarge" bold>
          Oops! ðŸ™ˆ
        </Text>

        <Text textAlign="center">
          NÃ£o foi possÃ­vel carregar os posts. Por favor, verifique sua conexÃ£o e
          tente novamente.
        </Text>
      </Box>

      <Button text="Tentar novamente" onPress={refetch} />
    </Box>
  );
}
