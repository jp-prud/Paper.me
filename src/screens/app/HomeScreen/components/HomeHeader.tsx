import {useNavigation} from '@react-navigation/native';
import {useGetUserData} from '@useCases';

import {Avatar, Box, Icon, Text, TouchableOpacityBox} from '@components';

export function HomeHeader() {
  const {userData, isLoading: userDataLoading} = useGetUserData();

  const {navigate} = useNavigation();

  function handleClickNavigateToProfileScreen() {
    navigate('ProfileScreen');
  }

  function handleClickNavigateToNotificationScreen() {
    navigate('SearchScreen');
  }

  return (
    <Box
      flexDirection="row"
      alignItems="center"
      justifyContent="space-between"
      pb="s12"
      borderBottomColor="gray4"
      borderBottomWidth={1}
      backgroundColor="background"
      testID="home-header">
      <TouchableOpacityBox
        onPress={handleClickNavigateToProfileScreen}
        testID="avatar-header-container">
        {!userDataLoading && <Avatar user={userData} />}
      </TouchableOpacityBox>

      <Text preset="headingSmall" bold>
        In√≠cio
      </Text>

      <TouchableOpacityBox
        width={32}
        height={32}
        justifyContent="center"
        alignItems="center"
        onPress={handleClickNavigateToNotificationScreen}
        testID="search-header-container">
        <Icon name="search" />
      </TouchableOpacityBox>
    </Box>
  );
}
