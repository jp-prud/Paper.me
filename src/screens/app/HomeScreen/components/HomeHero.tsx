import {Pressable, Image} from 'react-native';

import {Box, Icon, Text} from '@components';

export function HomeHero() {
  return (
    <Pressable>
      <Box flexDirection="column" gap="s16">
        <Box height={180} overflow="hidden" borderRadius="s16" flex={1}>
          <Image
            source={require('../../../../assets/images/bannerHome.png')}
            resizeMode="stretch"
            resizeMethod="scale"
          />
        </Box>

        <Box gap="s14">
          <Text preset="headingSmall">
            Saiba quais s√£o as vantagens de criar uma rotina organizada
          </Text>

          <Box flexDirection="row" alignItems="center">
            <Text color="neutral500" semiBold>
              Lifestyle
            </Text>

            <Box
              flex={1}
              justifyContent="flex-end"
              flexDirection="row"
              gap="s20">
              <Box flexDirection="row" alignItems="center" gap="s8">
                <Icon name="search" color="neutral500" size={24} />
                <Text color="neutral500" semiBold>
                  1h ago
                </Text>
              </Box>

              <Box flexDirection="row" alignItems="center" gap="s8">
                <Icon name="message" color="neutral500" />
                <Text color="neutral500" semiBold>
                  1
                </Text>
              </Box>
            </Box>
          </Box>
        </Box>
      </Box>
    </Pressable>
  );
}
