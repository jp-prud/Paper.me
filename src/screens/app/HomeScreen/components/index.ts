export * from './HomeHeader';
export * from './HomeErrorPostList';
export * from './HomeEmptyPostList';
