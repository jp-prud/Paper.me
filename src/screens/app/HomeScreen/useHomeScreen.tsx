import {usePostList} from '@domain';

export function useHomeScreen() {
  const {error, isLoading, postList, refetch} = usePostList();

  return {
    refetch,
    error,
    isLoading,
    postList,
  };
}
