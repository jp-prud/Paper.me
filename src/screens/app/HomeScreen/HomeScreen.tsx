import {FlatList, ListRenderItemInfo} from 'react-native';

import {PostProps} from '@domain';

import {Box, PostItem, RenderIfElse, Screen} from '@components';
import {AppScreenProps} from '@routes';

import {HomeHeader, HomeErrorPostList, HomeEmptyPostList} from './components';
import {useHomeScreen} from './useHomeScreen';

export function HomeScreen({}: AppScreenProps<'HomeScreen'>) {
  const {postList, isLoading, error, refetch} = useHomeScreen();

  function renderPost({item}: ListRenderItemInfo<PostProps>) {
    return <PostItem post={item} />;
  }

  function renderSeparator() {
    return <Box height={1} backgroundColor="gray4" mb="s24" mt="s24" />;
  }

  function renderPostList() {
    return (
      <FlatList
        data={postList}
        renderItem={renderPost}
        keyExtractor={item => item.id}
        showsVerticalScrollIndicator={false}
        ListHeaderComponent={<HomeHeader />}
        stickyHeaderIndices={[0]}
        ListEmptyComponent={<HomeEmptyPostList />}
        ItemSeparatorComponent={renderSeparator}
      />
    );
  }

  return (
    <Screen isLoading={isLoading}>
      <RenderIfElse
        condition={error}
        renderIf={<HomeErrorPostList refetch={refetch} />}
        renderElse={renderPostList()}
      />
    </Screen>
  );
}
