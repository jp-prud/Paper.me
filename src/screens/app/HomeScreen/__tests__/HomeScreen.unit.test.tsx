import {render, renderCustomScreenComponent} from '@tests';

import {HomeHeader} from '../components';
import {HomeScreen} from '../HomeScreen';

describe('HomeScreen - unit', () => {
  it('should render the componet', () => {
    const {toJSON} = render(
      <HomeScreen navigation={undefined} route={undefined} />,
    );

    expect(toJSON()).toMatchSnapshot();
  });

  it('should render HomeHeader when not isLoading and no error', () => {
    const {getByTestId} = renderCustomScreenComponent({
      customHeader: <HomeHeader />,
      children: <></>,
    });

    expect(getByTestId('home-header')).toBeTruthy();
  });
});
