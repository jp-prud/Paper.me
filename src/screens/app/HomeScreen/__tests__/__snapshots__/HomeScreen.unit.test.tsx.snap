// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomeScreen - unit should render the componet 1`] = `
[
  <View
    onLayout={[Function]}
    style={
      [
        {
          "backgroundColor": "#fefefe",
          "flex": 1,
        },
        {
          "paddingBottom": 0,
        },
      ]
    }
    testID="screen-component"
  >
    <View
      style={
        {
          "backgroundColor": "#fefefe",
          "flex": 1,
        }
      }
    >
      <View
        style={
          [
            {
              "flex": 1,
              "justifyContent": "center",
            },
            [
              {
                "paddingBottom": 20,
                "paddingHorizontal": 24,
                "paddingTop": 20,
              },
              undefined,
            ],
          ]
        }
      >
        <View
          style={
            [
              {
                "alignItems": "center",
                "justifyContent": "center",
              },
            ]
          }
        >
          <ActivityIndicator
            color="#017582"
            size={48}
            testID="activity-indicator-component"
          />
        </View>
      </View>
    </View>
  </View>,
  " ",
]
`;
